util/auto_tune_compressor.cc

```diff
@@ -144,138 +153,211 @@ CostAwareCompressor::CostAwareCompressor(const CompressionOptions& opts)
       continue;
     }
     if (kCompressionLevels[type - 1].size() == 0) {
-      allcompressors_.emplace_back();
       continue;
     } else {
-      // if the compression type is not supported, then skip and remove
-      // compression levels from the supported compression level list
+      // If the compression type is not supported, then skip and remove
+      // compression levels from the supported compression level list.
       if (std::find(compressions.begin(), compressions.end(), type) ==
           compressions.end()) {
-        allcompressors_.emplace_back();
         continue;
       }
-      std::vector<std::unique_ptr<Compressor>> compressors_diff_levels;
       for (size_t j = 0; j < kCompressionLevels[type - 1].size(); j++) {
         auto level = kCompressionLevels[type - 1][j];
         CompressionOptions new_opts = opts;
         new_opts.level = level;
-        compressors_diff_levels.push_back(
+        compressors_.emplace_back(
             builtInManager->GetCompressor(new_opts, type));
-        allcompressors_index_.emplace_back(i, j);
       }
-      allcompressors_.push_back(std::move(compressors_diff_levels));
     }
   }
+  MeasureUtilization();
+  block_count_ = 0;
+  cur_compressor_idx_ = 0;
 }
-
-const char* CostAwareCompressor::Name() const { return "CostAwareCompressor"; }
-size_t CostAwareCompressor::GetMaxSampleSizeIfWantDict(
-    CacheEntryRole block_type) const {
-  auto idx = allcompressors_index_.back();
-  return allcompressors_[idx.first][idx.second]->GetMaxSampleSizeIfWantDict(
-      block_type);
-}
-
-Slice CostAwareCompressor::GetSerializedDict() const {
-  auto idx = allcompressors_index_.back();
-  return allcompressors_[idx.first][idx.second]->GetSerializedDict();
+AutoCompressionAlgoLevelSelector::~AutoCompressionAlgoLevelSelector() {}
+const char* AutoCompressionAlgoLevelSelector::Name() const {
+  return "AutoCompressionAlgoLevelSelector";
 }
-
-CompressionType CostAwareCompressor::GetPreferredCompressionType() const {
-  return kZSTD;
-}
-std::unique_ptr<Compressor> CostAwareCompressor::MaybeCloneSpecialized(
+std::unique_ptr<Compressor>
+AutoCompressionAlgoLevelSelector::MaybeCloneSpecialized(
     CacheEntryRole block_type, DictSampleArgs&& dict_samples) {
   // TODO: full dictionary compression support. Currently this just falls
   // back on a non-multi compressor when asked to use a dictionary.
-  auto idx = allcompressors_index_.back();
-  return allcompressors_[idx.first][idx.second]->MaybeCloneSpecialized(
-      block_type, std::move(dict_samples));
+  assert(compressors_.size() > 0);
```

Comment:
Is `compressors_.size() > 0` guaranteed? Is it possible that  compressors_.emplace_back() never happens in the constructor? If that case, what will happen?